planning/motion_velocity_smoother/src/smoother/smoother_base.cpp

```diff
@@ -125,4 +166,71 @@ boost::optional<TrajectoryPoints> SmootherBase::applyLateralAccelerationFilter(
   return output;
 }
 
+boost::optional<TrajectoryPoints> SmootherBase::applySteeringRateLimit(
+  const TrajectoryPoints & input) const
+{
+  if (input.empty()) {
+    return boost::none;
+  }
+
+  if (input.size() < 3) {
+    return boost::optional<TrajectoryPoints>(
+      input);  // cannot calculate the desired velocity. do nothing.
+  }
+  // Interpolate with constant interval distance for lateral acceleration calculation.
+  std::vector<double> out_arclength;
+  const auto traj_length = motion_utils::calcArcLength(input);
+  for (double s = 0; s < traj_length; s += base_param_.sample_ds) {
+    out_arclength.push_back(s);
+  }
+  const auto output_traj =
+    motion_utils::resampleTrajectory(motion_utils::convertToTrajectory(input), out_arclength);
+  auto output = motion_utils::convertToTrajectoryPointArray(output_traj);
+  output.back() = input.back();  // keep the final speed.
+
+  const size_t idx_dist = static_cast<size_t>(std::max(
+    static_cast<int>((base_param_.curvature_calculation_distance) / base_param_.sample_ds), 1));
+
+  // Calculate curvature assuming the trajectory points interval is constant
+  const auto curvature_v = calcCurvatureFrom3Points(output, idx_dist);
```

Comment:
I have talked with @TakaHoribe.
He is concerned that curvature calculation in other parts of the code (e.g., https://github.com/autowarefoundation/autoware.universe/blob/3a7b9c2fcf0a6025735414fe71211bfd469fe822/planning/motion_velocity_smoother/src/smoother/smoother_base.cpp#L92).
Usually, just using adjacent points to calculate curvature creates very noisy curvature, and the code in trajectory_utils will use longer distance to calculate more "generic" curvature.
Do you think you can replace this function with the one in `trajectory_utils`?

Reply:
> https://github.com/autowarefoundation/autoware.universe/blob/3a7b9c2fcf0a6025735414fe71211bfd469fe822/planning/motion_velocity_smoother/src/smoother/smoother_base.cpp#L92

> Usually, just using adjacent points to calculate curvature creates very noisy curvature, and the code in trajectory_utils will use longer distance to calculate more "generic" curvature.

Actually, I used this function first but it copied first and last curvature points (that can not calculate due to distance)  from next or before points that can be calculated so it causes curvature with high error. I think noisy curvature should be better than wrong calculated curvature.

@TakaHoribe what do you think about this?

Reply:
@brkay54 You mean, [this part](https://github.com/autowarefoundation/autoware.universe/blob/main/planning/motion_velocity_smoother/src/trajectory_utils.cpp/#L226-L231)?

The curvature computation with 3 adjacent points is too noisy for practical use. The function in `trajectory_utils` calculates the curvature from 3 distant points. The drawback is that it doesn't work around the beginning and the end of the path. To prevent that, the path has a distance behind the ego.

The curvature calculation proposed in this PR is different from the existing method, but essentially it does the same thing. So the concern remains with smoothing. Let's look at the results of the experiment. If the method proposed in this PR is better, then it is nice to think about replacing the calculation method used in the lateral acceleration filter as well.

Reply:
> 



> @brkay54 You mean, [this part](https://github.com/autowarefoundation/autoware.universe/blob/main/planning/motion_velocity_smoother/src/trajectory_utils.cpp/#L226-L231)?

Yes.

> 
> The curvature computation with 3 adjacent points is too noisy for practical use. The function in `trajectory_utils` calculates the curvature from 3 distant points. The drawback is that it doesn't work around the beginning and the end of the path. To prevent that, the path has a distance behind the ego.

Yes almost same thing, [here](https://github.com/autowarefoundation/autoware.universe/pull/1071/files#diff-299b94dd4370e4b0fd7897c78db0eccc8818748ef19b9e0c00b717bcf878b851R89-R99) for the beginning of the trajectory, let say idx_distance is 5. For first point (idx = 0) curvature is zero. 

Lets look at first indexes of path and which trajectory points will be used to calculate curvature.
For idx = 1 -> [0, 1, 6]
For idx = 2 -> [0, 2, 7] 
For idx = 3 -> [0, 3, 8]
...
For idx = 6 -> [1, 6, 11]

So it not use adjacent points for all the path just for beginning and end of the path.

Reply:
I plotted the interpolated curvature results of these two approach. I tested them in a sharp turn in gebze_map.

Reply:
@brkay54 Thank you! Your approach looks nice and I don't see any drawbacks. Would you make a separate PR to update the existing `trajectory_utils::calcTrajectoryCurvatureFrom3Points()` function with this method? I'd like to make it a default.

Reply:
New approach is merged with https://github.com/autowarefoundation/autoware.universe/pull/1638